### ISSUE TITLE
WebGLRenderer: Add support for Node Materials

### ISSUE THREAD

**Issue by gkjohnson** [mrdoob/three.js#30185]
### Description

I wanted to open a discussion to talk about the possibility of adding Node Materials to WebGLRenderer - as far as I understand Node Materials are only supported with WebGPURenderer which poses a lot of issues for community adoption, practical testing, and library support:

WebGPURenderer looks like it's coming along great but until it's more fully featured and mature I cannot use it as a platform for any of my professional work. Likewise many tools I've developed and work on are already built on WebGLRenderer and cannot ported to WebGPURenderer. I assume many developers are in the same boat. This means I haven't been able to invest any time in learning and trying the new node material system despite being very interested in the benefits after years of trying to solve shader and material problems.

Adding support to WebGLRenderer would help afford the well-experienced three.js community the ability to test the system in large scale, complex projects and give practical feedback sooner rather than trying it to WebGPURenderer. This will also help ease the transition projects and developers that currently need to use WebGLRenderer. Currently all material development currently being done for WebGL will be rendered useless in WebGPU which this can help avoid.

And from the library perspective - I need to be able to develop something that will work in both WebGLRenderer and WebGPURenderer. Asking developers to maintain two versions of their materials or libraries isn't great and I think will otherwise just make this renderer transition significantly more difficult than it needs to be.

I appreciate all the work done on the new material systems - it looks like a large improvement over what we have now so it would be great to make this transition as smooth as possible.

cc @sunag @RenaudRohlinger 

### Solution

Add support for NodeMaterials to WebGLRenderer.

### Alternatives

Community feedback is significantly delayed, developers and libraries are fragmented.

**Comment by DennisSmolek** [comment]
Correct me if I'm wrong, but doesn't switching to the WebGPURenderer (which supports nodes) work on WebGL only instances as the fallback?

Or are you wanting a way to mix the existing GLSL code with the outputs of the node system?

Are you just wanting the output of the visual side of things or are you wanting things like the compute and functional nodes as well?

**Comment by gkjohnson** [comment]
I understand that WebGPURenderer has a webgl fallback but this is about using Node Materials with WebGLRenderer. I don't care about the underlying browser graphics API being used - I care about the compatibility with the WebGLRenderer class that many projects are heavily invested in already and cannot or are not ready to switch for a variety of reasons.

> Or are you wanting a way to mix the existing GLSL code with the outputs of the node system?

I just want to use node materials as they are with WebGLRenderer and leverage all the benefits that the node system already promises. I'm not proposing new features otherwise.

**Comment by sunag** [comment]
Thanks for the analysis @gkjohnson, it's a very interesting subject. In these last months where WebGPURenderer is gaining more users, it became clear that we need a migration guide to TSL, not just the a specification. I believe that just as it was done in previous years to design the new Nodes and TSL system, now it will be to smooth the transition.

I fear that it will not be possible to migrate all the TSL features, except for the basic functions already known in the old node system, but I believe that this is not enough. We had to discontinue the new node system for WebGLRenderer because it was falling far behind in features compared to WebGPURenderer (https://github.com/mrdoob/three.js/pull/28167), the renderer ended up being re-architected not only because of WebGPU but also to align with the new node system, and bring new features like `backdrop` for example, things that would not be possible using GLSL with simple functions in WebGLRenderer without complex work involved, only in this specific case WebGPURenderer does not render the opaque materials twice when using transmission materials like WebGLRenderer, instead it copies the g-buffer to a texture, for this it is also necessary that the color transform is done in post-production internally, this backdrop effect can be accessed on any material using `viewportTexture()`. The event system like '.updateBefore()' that allows nodes to create pre-pass, at the time of rendering the object, which can also be shared with other related ones, in WebGPURenderer the light and shadow system can be extended using nodes, and this allows users create new shadows and shadow filters as well as the light, light model and the light system without needing to modify the core for this. Including other things like `compute()`... surely this is not all just what I remembered now.

Alternatively, I would like to know if there is anything else we can improve like the Three.js Transpiler, and a guide for this migration.

**Comment by Mugen87** [comment]
I'd like to understand in more detail why migrating to the new renderer is any issue. To me, the real migration tasks is to mpve from GSLS to TSL. We want to support this process with more documentation and guides and also technically with a TSL transpiler. But `WebGPURenderer` itself is not that different compared to `WebGLRenderer`, at least API-wise. Most public methods are identical, it is required in some cases to switch to a async/await pattern though. As long as you didn't use an extensively forked version of the renderer, just exchanging the renderers should not create a lot of migration task.

Generally speaking, TSL and `WebGPURenderer` are far superior compared to `WebGLRenderer` with its hardwired material and lights system. As outlined by @sunag supporting node materials in `WebGLRenderer` is essentially developing an new `WebGLRenderer`. I've said it earlier but the project should not make the mistake to invest too many resources in maintaining things which are in fact legacy. From a strategic point of view, it's better to focus on the future and the already existing and advances alternatives. I understand when companies (or individuals) have built solutions around `WebGLRenderer` but it should be clear that a migration to TSL and `WebGPURenderer` is inevitable. I understand there are some reservations about the maturity level of WebGPU but it's also important to highlight that a WebGPU usage isn't mandatory when using the new renderer. WebGL 2 can always be forced with a renderer parameter.

TBH, I have expected this discussion sooner or later since we always had it when moving to new approaches. E.g. `Geometry` to `BufferGeometry`, WebGL 1 to WebGL 2 or legacy to physically correct lighting. There were always developers who wanted to retain the old structures for different reasons but the project consistently deprecated (and eventually removed) legacy code paths. Which was in retrospective the right decision. It should be clear that at some point in the future `WebGLRenderer` will be deprecated as well. To me, the renderer is now in some sort of pre-stage. We still provide bug fixes and smaller additions, but implement no big features or larger refactorings anymore. And I think that's the right direction.

**Comment by gkjohnson** [comment]
Thanks Sunag - I understand the new nodes system includes a lot of "magic" that can't necessarily be easily back ported to WebGLRenderer without significant work and I think that's okay. A more standard subset of features akin to what would have been supported previously or with a more traditional nodes system would be sufficient imo and serve as an on-ramp for users - then more complex nodes and TSL systems can be "better with WebGPURenderer".

I'm mostly looking for a way to make sure (some very complex) materials and shaders being written right now can be used later down the line or a porting process can be done progressively rather than all at once, possibly fragmenting the community in the process (libraries will have to decide to support their huge userbase of WebGLRenderer users, abandon them and only support WebGPURenderer, or support two completely different codebases).

---

Mugen -

> E.g. Geometry to BufferGeometry, WebGL 1 to WebGL 2 or legacy to physically correct lighting.

I'm having a hard time seeing how these changes can be compared to upgrading to WebGPURenderer at all. WebGL1 > 2 was transparent for almost everyone since it was enabled by default and physically correct lighting was a flag and multiplying light intensities by some constants. Deprecating Geometry is maybe the only significant one but even when I started using the project almost 10 years ago, using "Geometry" already did not feel recommended. There have been a lot of deprecations and whatnot over the lifespan of three.js and for what it's worth none of them have really been concerning. I think it needs to be recognized how much more significant this change is.

> Most public methods are identical, it is required in some cases to switch to an async/await pattern though

This alone can mean _significant_ architectural changes for user projects, guarding against race conditions, etc. 

> I also understand there are some reservations about the maturity level of WebGPU but it's important to highlight that a WebGPU usage isn't mandatory

To be clear I wasn't expressing concern about the underlying webgpu browser API - I'm expressing concern about the three.js WebGPURenderer implementation. WebGLRenderer has over 10 years of bug fixes, API refinement, memory leak fixes, etc. Of course a lot's been learned in that time and is applied to WebGPURenderer but I've written and used enough code to know that a complete rewrite of anything needs some time to be proven before it should be relied on for any large scale projects.

Dependencies on ecosystem libraries is also something that needs to be considered. Tens of thousands of projects using the "postprocessing" package _every week_, for example, and it will not work with WebGPURenderer which means any project that depends on it will not be able to upgrade. I know this project has post processing effects, as well, but historically the three.js screen effects have not been comparable in terms of performance or quality - which is fine. It should be expected that dedicated package is higher quality.

That project has been in the process of upgrading its architecture for [almost 4 years](https://github.com/pmndrs/postprocessing/issues/279). How long until WebGPURenderer will be supported? Again no knock on the project; it's great work and we all know this stuff takes time especially when it's free contributions. But the reality is that many people rely on projects like this and it's going to impact the sense of "readiness" and ability to migrate to a newer three.js renderer.

I in-part created this issue because another package, [three-geospatial](https://github.com/takram-design-engineering/three-geospatial/), is creating some amazing, complex effects and because of the above limitations the codebase is stuck in WebGLRenderer and I foresee the maintainer not having the time, availability, or interest in porting it forward. Any project depending on this one now will not be able to migrate forward now, either. This may be a small aside but the notation of TSL also makes writing the scale of shaders such as the ones in that project feel completely unapproachable, which is one of the areas I'm hoping more feedback can be surfaced.

Three.js should be promoting building modern, future-proof shaders which should ultimately help accelerate us to a WebGPURenderer-only world rather than continuing to have to build using the old system simply because dependencies and renderer maturity are preventing it. I feel like the inability to write _any_ cross-renderer shaders is already having a larger impact than people seem to be aware of. I'm curious to hear the thoughts on the above issues.

To be clear I'm very excited for WebGPURenderer and the future of node materials. To be honest I've been excited for node materials for the last several years and truly I'd be hoping that I'd have been able to take more practical advantage of them by now 😭

**Comment by vanruesc** [comment]
> That project has been in the process of upgrading its architecture for almost 4 years.

Has it really been that long already :skull: 

> How long until WebGPURenderer will be supported?

I'm excited for node materials but haven't used them for the same reasons you've described. If WebGPURenderer supports shader-based materials alongside node materials, it should be a drop-in replacement for `postprocessing`.

The part that would need to be changed in `postprocessing` is the manipulation of three's built-in materials to render additional scene data during the main geometry pass via MRT. From my point of view, node materials just provide a better way of manipulating those built-in materials and while I've been curious about that aspect in the past (https://github.com/mrdoob/three.js/issues/16440#issuecomment-632259369, https://github.com/mrdoob/three.js/issues/16440#issuecomment-632671252) I've not had the chance yet to actually try it out.

WebGLRenderer's materials can only be tweaked via `onBeforeCompile` string manipulation which is fragile. A while ago, I figured out that those manipulations can largely be avoided for the MRT use-case if `three` would define explicit and predictable shader outputs (https://github.com/mrdoob/three.js/pull/27808). And I think node materials already provide something similar. However, WebGLRenderer is still widely used, so I'll try to get back to finishing that PR because it would make the switch to node materials easier later on.

**Comment by CodyJasonBennett** [comment]
If it helps, another perspective is that the use of `WebGPURenderer` requires a complete rewrite for the surrounding ecosystem, and there's no clear path to support both `WebGLRenderer` and `WebGPURenderer`, as the ecosystem will inevitably be fractured. The only present solution requires multiple builds and a bespoke shader chunk system using TSL transpilers or forked implementations, which is simply infeasible. Further, as WebGPU and WGSL are not fully specified, we can't productionize on them at all in the browser, which personally limits me to contributing as an enthusiast. Backporting node materials would serve as this path and a platform to write future-proof shaders today that leverage new backends later.

This is incidentally nice for `postprocessing` since it doesn't have to fork from the uber material system to create a G-buffer necessary for material properties like roughness and normals (maps) for SSR/GI. [My SSILVB](https://x.com/Cody_J_Bennett/status/1815475933744374032) based on Activision GTAO did for normal map detail, but I had since abandoned it due to a plethora of ANGLE/WebGL bugs. Original plans and linked triage were from when nodes were WebGL only (no TSL) and `WebGLRenderer` did support it. I maintained this original system and addons in the `three-stdlib` fork and here, but I since removed it in v2.

I would be happy to champion this work if the project allows, as my work was prepared to drop use of three upstream for our own production renderers (OSS, on top of three) since we felt there was an ivory tower status around future development in three.js and had a bigger stake in projects like Slang over TSL. If we could open up discussion on nodes and agree on some goals, we have much to contribute that would otherwise be potentially duplicate. This includes graphics work, which I have been careful to backport or mirror when possible.

In any case, I plan to upstream fixes or features if able, but my general ambition is greatly limited without clear knowledge as to what is in flight and what use cases three.js looks to with these systems. This is a discussion I've been meaning to have since 2021 as nodes materialized in the `WebGPURenderer`, but opted to play and contribute to specs directly instead. My focus was instead to WebGL2, which Renaud has continued in open source.

**Comment by mrdoob** [comment]
@sunag How hard would it be to implement something along the lines of this?

```js
import * as THREE from 'three';
import { GLSLNodeBuilder, NodeMaterial } from 'three/webgpu';
import { color, float, time, sin, mul, add } from 'three/tsl';

const renderer = new THREE.WebGLRenderer();
document.body.appendChild(renderer.domElement);

const builder = new GLSLNodeBuilder(); // <--

const timeFactor = sin(mul(time, float(2.0)));
const dynamicColor = mul(color(0.5, 0.2, 1.0), timeFactor);

const nodeMaterial = new NodeMaterial();
nodeMaterial.colorNode = dynamicColor;

const material = builder.build(nodeMaterial); // <--

const geometry = new THREE.SphereGeometry(1, 32, 32);
const mesh = new THREE.Mesh(geometry, material);
scene.add(mesh);

function animate() {
  requestAnimationFrame(animate);
  renderer.render(scene, camera);
}
animate();
```

**Comment by CodyJasonBennett** [comment]
It seems that nodes have a deep dependency on the current renderer/backend (setMRT/nodes/library are entrenched), but what MrDoob suggested I think is a good goal. I don't see a way to completely backport support without backporting the whole rendering API with unconditional use of nodes. Alternatively, #29781, but this could be done in a future-proof way with TSL, which reduces that use-case to portability.

**Comment by sunag** [comment]
For `WebGLRenderer` I think the only viable way is to bring `WebGLNodeBuilder` back. `GLSLNodeBuilder` will produce very different code that is incompatible with `ShaderChunk` and is based on `UniformGroups` for all uniforms but I still wonder if we should invest in a legacy renderer as @Mugen87 said https://github.com/mrdoob/three.js/issues/30185#issuecomment-2559314704... @brunosimon ( https://www.youtube.com/watch?v=C2gDL9Qk_vo ), @RenaudRohlinger ( https://www.utsubo.com/ ) and others have shown that it is already possible to make amazing projects in `WebGPURenderer`, it is fair to be afraid because it is new and we may encounter bugs along the way, but the more projects are made in `WebGPURenderer` the sooner we can polish the renderer and make it more stable and performant, and your help would be amazing. I also have doubts about the instability that the node system can have in `WebGLRenderer`.

With `WebGPURenderer` we already have MRT support built into materials and post-processing, so it could be great to spend our time trying to make these features compatible with the new renderer. The result of this test is more related to the use of the approach than to the backend: https://github.com/mrdoob/three.js/pull/29597#issuecomment-2402597482. I know we don't have complete documentation yet, but we are open to help in any way we can, creating features, answering questions, creating examples and improving the system to speed up this development.

TSL is also not a "closed language", before being a shader syntax it is a Three.js graphics API, so it is also possible to port WGSL and GLSL functions, if the focus is programming for a specific backend, `wgslFn` and `glslFn` nodes can be used and we can probably improve this support as well. A lot of the functionality we have exclusive to `WebGPURenderer` comes from users who were using the renderer as a base for their projects and were looking for WebGPU functionality, and we did that, with most of them falling back to WebGL2. So to me the alternative mentioned https://github.com/mrdoob/three.js/issues/29781 seems to make more sense for me.

**Comment by Mugen87** [comment]
Adding support of TSL will also prolong the lifetime of `WebGLRenderer` and constantly produce new feature requests so a big amount of our time won't be invested in `WebGPURenderer` anymore. We already support occlusion queries or transform feedback in `WebGPURenderer`'s WebGL backend and I can guarantee the community wants these for `WebGLRenderer` as well if now include TSL support. This is just the wrong signal and I do not support going down that path.

It's like when a car manufacture would now invest budget in enhancing a fuel-based combustion engine instead of investing in electric engines, battery tech and autonomous driving. There are better alternatives so let's focus on them.

And to clarify: It is not sufficient to just generate the shader source from TSL. As @sunag mentioned the whole uniforms/bindings topic is implemented differently (more efficiently) in `WebGPURenderer` so we would need a complete separate code path (`WebGLNodeBuilder`) for `WebGLRenderer`.

**Comment by CodyJasonBennett** [comment]
Is it technically possible to implement the backend interface in `WebGLRenderer` so it would align with `WebGLBackend`? I understand there's still a strict dependency on nodes for library/mrt fields, and node generators have a deep dependency on the current backend; otherwise, it can be used in `WebGLRenderer` all the same. Ideally, generators would not have a dependency on backends at all and reference a capabilities or pipeline layout object. I ask since I'm willing to do this work myself and would pay just for the ability to do it, even if limited to a subset of features (e.g., no transform feedback). The alternative #29781 still puts the ecosystem in stasis like [pmndrs/postprocessing](https://github.com/pmndrs/postprocessing), and it's uncomfortable to do this in user-land since TSL is not a real standard, so we're beholden on a patch basis to three.js (the only implementation) all the same. In any case, I'm concerned about shutting down any stepping stone equal to any migration path.

**Comment by Makio64** [comment]
Yesterday, while talking about this subject, I was wondering if the solution would be to add a Legacy version into the WebGPURenderer.

For example : `MeshBasicMaterialLegacy ` with the same system than we had before ( chunks, onbeofrecompile etc.. ) but in the end the fragment & vertex shaders are push into a glslFn(...). 
That way old shader & material can be transfert easily to WebGPURenderer and benefit from the new updates while not having to translate all of them into tsl ?

Of course this is only to simplify the transition to WebGPURenderer for old project which cant spend months translating all..

**Comment by nkallen** [comment]
Just thought I’d put my two cents in. Plasticity is a CAD program that uses the legacy webgl renderer. Our code is performance sensitive (we have large scenes and must use a fork of BatchedMesh) and we are likely too risk averse to use the new fallback renderer. (We will switch to webgpu when it is mature, perhaps in early 2026). 

I just had someone implement triplanar and cylindrical mapping using the old shaderchunk system. It looks freaking awesome. My first thought is: let’s contribute this back! But it’s my understanding this feature cannot be contributed back to threejs because shaderchunks are under a de facto code freeze. Since the node system cannot be used with the legacy renderer and the decision has been made not to support nodes on the legacy renderer, and webgpu isn’t really ready for prime time in the next year … this situation does seem like a pickle. 

My instincts as a software engineer are telling me that back porting nodes is a desirable, conservative, incremental upgrade path. It allows the community to benefit from nodes (eg materialx) as well as make enhancements to them under the more widespread usage of the wider legacy user base. 

I understand this is a volunteer project and this is also one reason why companies like us donate money (and are willing to donate more) so that the best decisions can be made.

**Comment by mrdoob** [comment]
> Since the node system cannot be used with the legacy renderer and the decision has been made not to support nodes on the legacy renderer, and webgpu isn’t really ready for prime time in the next year … this situation does seem like a pickle.

As far as I know, no decision has been made on that topic yet. But it's true that the main contributors have been protective about taking on too much work and instead they've tried to push for the solutions they're focusing on.

In my opinion, as I just wrote in https://github.com/mrdoob/three.js/pull/30695#discussion_r2018629761, I think some sort of `LegacyShaderMaterial` and getting TSL working in `WebGLRenderer` should be much more viable now that we can leverage AI assisted development.

However, I feel like it's hard to work on both solutions at the same time... If I had to prioritize I think being able to bring your legacy shader materials to `WebGPURenderer` should be done first. But I understand that being able to use TSL in `WebGLRenderer` is equally important. At the same time, we have to add support for all the new features WebGPU is implementing and that's what @sunag is focusing on.

At the end of the day, we mostly have a lack of resources and the few we have are being protective with their time.
Any help that doesn't create more work for us would be highly appreciated 🙏

**Comment by nkallen** [comment]
AI will for sure help, but I am less optimistic about it being a huge boon. More like 20% for this sort of tricky work I would guess.... All that said, Plasticity will strongly consider sponsoring enhancements to threejs that benefit us. We would benefit from TSL working in the WebGLRenderer so I wouldn't hesitate to spend $10k on it. If you need more I need to think about it.

**Comment by mrdoob** [comment]
I can do some explorations but I can't focus on it at this point.
Would @CodyJasonBennett be up for it for $10k?

**Comment by sunag** [comment]
> For WebGLRenderer I think the only viable way is to bring WebGLNodeBuilder back. 

It is very likely that if we revert [this PR](https://github.com/mrdoob/three.js/pull/28167) and update according to new changes and paths the basic features will work.

**Comment by mrdoob** [comment]
I think it's worth a try! Would you like to give it a go?

**Comment by RenaudRohlinger** [comment]
I've also spent a lot of time working on the current `GLSLNodeBuilder` and `WebGLBackend`, so I'd be happy to assist @sunag with this task as well. I'm always glad to contribute more to Three.js development, especially if there's an opportunity to be compensated for the work.

**Comment by nkallen** [comment]
Awesome. I think the two projects that mrdoob is proposing make a lot of sense and I'm happy to sponsor one of them. The reality is webgpu deployment is going to take a long time and webgl will be around for a long time. The web is the world's largest distributed system and the deployment strategy reflect that-- slow rollout, easy migration path, no feature freeze on legacy.

**Comment by sunag** [comment]
> I've also spent a lot of time working on the current GLSLNodeBuilder and WebGLBackend, so I'd be happy to assist @sunag with this task as well.

Yes, @RenaudRohlinger has been doing an incredible job! Sharing the effort would benefit the community even more. As @mrdoob mentioned, my focus is on developing new features and improving `WebGPU` support for `WebGPURenderer`, so having @RenaudRohlinger dedicate more time especially with sponsorship, would be a great advantage.

@nkallen Thanks!

**Comment by jellychen** [comment]
![Image](https://github.com/user-attachments/assets/dbec35f3-ad68-4da4-b7aa-614413302ce6)

We are developing a browser-based 3D modeling software similar to Blender, with Three.js as the rendering engine. This project has been under development for over three years, and Three.js has provided tremendous convenience, allowing us to focus on geometric algorithms while leveraging its capabilities for rendering and scene management. @mrdoob @sunag @gkjohnson 

Recently, we've been planning to build a node system and have seen the potential of TSL Node. However, our previous system heavily relied on WebGLRenderer as the rendering engine, with many custom GLSL materials. This has created significant obstacles in our transition to WebGPU. In fact, we've already attempted the migration for some time but were forced to abandon it.

We have to carefully evaluate the cost of this transition. If Three.js could make TSL Node generate materials that are compatible with WebGL Renderer, it would be a tremendous help for our migration.

I believe many developers share the same dilemma as us. This issue is preventing us from adopting the node-based material system. If we were starting a new project from scratch, we might directly embrace the WebGPU ecosystem—but that's practically impossible for us right now.

**Comment by jellychen** [comment]
@mrdoob @sunag @gkjohnson If a better solution emerges, we'd be willing to contribute to the Three.js ecosystem—for example, by open-sourcing our node-based material editor. As a potential commercial project, we're committed to maintaining it long-term. If possible, we'd even hope for it to become one of Three.js's official editor options.

**Comment by jellychen** [comment]
With controllability in mind, we are now implementing the node system ourselves and then compatible with webgl renderer. But for the future of the project, it is definitely not compatible with TSL. This kind of gap is very disturbing

**Comment by RenaudRohlinger** [comment]
Looks like interest in bringing Node Materials back to the `WebGLRenderer` is growing! Should I go ahead and start reviving the PR, and gradually expanding support? @nkallen @sunag

To clarify the scope: the idea would be to restore the legacy WebGL Node Material system and extend it to cover all built-in materials used by the `WebGLRenderer`. I’d also ensure compatibility with the `MaterialXLoader`, with a priority on what’s needed for Plasticity, since they’re actively contributing to this effort.

Some newer nodes like `geometryNode`, `backdropNode`, `mrtNode` probably won’t be supported, and certain shadow-related nodes might behave slightly differently, as they rely on features specific to the WebGPURenderer. Users needing those would still need to move on to the new Renderer.

If @nkallen gives the go-ahead and other collaborators are aligned, I’m happy to begin next week.

**Comment by nkallen** [comment]
Our primary (short-term) goal is to keep our threejs fork as minimal as possible (or eliminate it entirely). Our secondary (long-term) goal is to make it as easy as possible for us to switch to webgpu when the browser support is mature enough. So for example, we recently added triplanar mapping to to the "old" shader system; it was too many files and too fragile to do with `onBeforeCompile` and so on. As far as I'm aware, TSL already supports triplanar mapping, and if it didn't or we wanted to make enhancements to it, you would be more likely to accept our pull-request if done in TSL. Thus, for us, once TSL in the old renderer, it reduces the size of our fork by 50%, which saves us a lot of time when we want to upgrade three.js. Then, slowly but surely, we can slowly port all of our custom shaders (zebra stripes, isophotes, draft shader, section hatching, etc) to TSL and the eventual migration to webgpu will be much easier/more incremental. I think it will also benefit the community to have a wider user-base "stress testing" the TSL shaders without having to go "full webgpu" since relatively few people can do that until browser support is better.

MaterialX is nice to have but not a necessity at the moment. We have no need for geometry transformations. I would say a good benchmark for us would just be to demonstrate the equivalent of meshphysicalmaterial using triplanar mapping in tsl on the old renderer. So yes, you have my go-ahead

**Comment by jellychen** [comment]
If the official intends to implement it, I should use your implementation to prevent too much difference. If this route is determined. What contributions can I make? @RenaudRohlinger @mrdoob

**Comment by nkallen** [comment]
Hey everybody - wanted to check in. Any progress on this? Any blockers?

**Comment by mrdoob** [comment]
I think @RenaudRohlinger has been busy with the Osaka Expo project.

**Comment by RenaudRohlinger** [comment]
Sorry guys I've been indeed pretty busy with work and the Osaka Expo this year, I think I should have some availabilities back around September.

**Comment by CodyJasonBennett** [comment]
I did propose championing the work myself, and I have already shared early exploration between modularizing individual renderer systems for nodes and more simply implementing the abstract `Renderer` interface, along with previous analysis for tree-shaking in add-ons. Should I take this as a green light for me to continue? The goal is to get even a cursory direction for how this can be done without diverging API on either end, and I do think discovery is necessary here. I won't commit to such work if the project won't accept them.

**Comment by sunag** [comment]
I still think the best way would be to revert [this PR](https://github.com/mrdoob/three.js/pull/28167) and update according to the changes, there are certainly deviations that I am not calculating here, but I will give some information that may be useful.

The `WebGLNodeBuilder` was responsible for generating the node system for the `WebGLRenderer`. Basically, it copies the material’s `THREE.ShaderLib` and injects the code generated by the `WebGLNodeBuilder`. A system called `SlotNode` was created to automate this replacement process, which was very useful for maintaining the code when the core shader was changed.

https://github.com/mrdoob/three.js/blob/0ecd4821355f623218e09957a18fc644cb071212/examples/jsm/renderers/webgl-legacy/nodes/WebGLNodeBuilder.js#L125-L130

It’s very likely that most features will continue to work, such as uniforms (floats, vectors, matrices, textures), arithmetic operators, automatic type conversions, maths, as well as some of the system optimizations. Some more advanced optimizations, like `subBuild()`—which automatically declares normals based on the NodeMaterial input—are more complex and may require the `WebGLNodeBuilder` to handle some specific optimizations, since `THREE.ShaderLib` defines many essential variables directly, such as `tnb` for `TBNViewMatrix`, `transformedNormal` for `normalView`, and so on.

`WebGLNodes` is responsible for updating the values—mainly uniforms—and passing them to the ShaderMaterial uniforms using `Material.onBeforeRenderer()`. `Material.onBuild()` was equivalent to `Material.onBeforeCompile()`, handling the automatic compilation of the NodeMaterial at first render, but it has been removed.

Some features features won’t be available, as I detailed [here](https://github.com/mrdoob/three.js/issues/30185#issuecomment-2558853997), because the internal architecture of the `WebGLRenderer` limits this. However, for the essential shader-graph features needed for material customization (and not for the entire lighting or compute model), it should still meet most needs.

**Comment by CodyJasonBennett** [comment]
I am not sure. It's tempting to look for an easy or quick segment of work here (many words on this), but I would be inclined to reference the existing `GLSLNodeBuilder`, although it is much more work. The motivation is to facilitate cross-referencing as it is maintained in time and open the door to widen compatibility. The shared API that MrDoob proposed would be ideal. It's my hope we can settle on an interface with less duplication or that can be shared. Currently, it is not easily possible to decouple renderer state from `NodeBuilder` (the purpose of my explorations), which is unfortunate, as I anticipate duplication needed rather than merely supporting and using the interface directly. I don't think legacy shader chunks should be used here at all if we can aim to support existing node materials.

**Comment by sunag** [comment]
@CodyJasonBennett This certainly wouldn't work without making a lot of modifications to the `WebGLRenderer` since as mentioned, several native TSL features are only available because the renderer architecture has changed. So `ShaderLib` is more viable anyway, but if you go the other way you'll probably have to make `WebGLRenderer` more like what `WebGPURenderer` is today.

**Comment by Mugen87** [comment]
> So ShaderLib is more viable anyway, but if you go the other way you'll probably have to make WebGLRenderer more like what WebGPURenderer is today.

When I understand the motivation of this issue correctly (especially @gkjohnson's initial post), having node support in `WebGLRenderer` is wanted because certain third-party libraries or external/custom tools do not work with `WebGPURenderer` yet. However, a more in-depth rewrite of `WebGLRenderer` will break such tools because they often rely on certain internals of `WebGLRenderer` (like `ShaderLib`).

To me, what @CodyJasonBennett suggest in https://github.com/mrdoob/three.js/issues/30185#issuecomment-3042610848 is contradictory to the idea to integrate node support into `WebGLRenderer` without changing its fundamental design.

If you would change `WebGLRenderer` to the same principles like the WebGL backend of `WebGPURenderer`, why not using `WebGPURenderer` in the first place?

That aside: I want to make clear again that don't think we should go for a future with more than one (official) renderer. In the long term, there should only be a single renderer with multiple backends and `WebGLRenderer` should be deprecated at some point. To me, it's not manageable for the project to work on two distinct renderers. 

There is no official timeline but let's assume `WebGLRenderer` would be deprecated in a few years. Instead of deprecating it, could it be maintained in a different repository by developers who still rely on it?

**Comment by Makio64** [comment]
@Mugen87 I think the naming is the root cause of some frustration here

Its expected to have breaking change when a famous package have a big update `vue2 -> vue3` or `vite v1->v2` etc.. and the lib around can say `support vue2 only` the fact there is no `threejs v2` make it hard for third-party libraries to communicate with their community. Related #29899

Make WebGPURenderer Friendly old libs / large project is a huge job and the initial request is trying to find a simple way to do it step by step. 

**I think we should do the opposite :** 
-> make a version of webgpurenderer compatible with shaderlibs ( including all shaderlibs old way ) -> compile the glsl -> transpile to tsl -> work done. ( probably there will be still other things to adjust but it seems to me more reasonable ) and then the libs / project can change to native tsl one shader at a time.

**Comment by nkallen** [comment]
The idiosyncratic things that we (Plasticity) find particularly valuable about this project:

1. Easy/gradual transition to webgpu, by first converting all of our material shaders to tsl, then later swapping to the webgpu renderer when it's stable/widely deployed
2. Access "now" to newer/advanced material shaders such as triplanar mapping without having to go full web-gpu
3. Wider community contribution to tsl shaders, since you have a wider userbase to test/enhance. (And note there is a de facto code freeze on legacy gsl code)

Things we (idiosyncratically/for our own selfish interests) would like to avoid
1. Destabilizing refactorings to the legacy renderer that might introduce bugs, performance degradations, etc.

==

I see @Makio64 just now proposed an alternative approach which is to compile legacy glsl to tsl. If it were possible, that does satisfy some of the above goals (easy transition to webgpu). Although it's probably not our #1 choice, it isn't a crazy idea if it is in fact easy (but is it??).

**Comment by Mugen87** [comment]
I can understand the wish for making the transition to `WebGPURenderer` easier. But I clearly favor @nkallen's approach. That could be achieved by what @sunag suggested in https://github.com/mrdoob/three.js/issues/30185#issuecomment-3042328316 via restoring `WebGLNodeBuilder`. That would allow to move to TSL first and then swap to `WebGPURenderer` at a later point.

The `ShaderLib` and shader chunk based approach is "dead" in my mind. It only exists in the space of `WebGLRenderer` and really shouldn't be a part of the new renderer.

**Comment by Makio64** [comment]
@Mugen87 `ShaderLib` and maybe `ShaderLibMaterial` should be an `addons` import, of course not part of the main project.

If you think about it, its a low ressource, fast & easy approach as you guys already build the `transpiler` to `tsl` : _we build a shader using old way > we transpile > done_

**Comment by nkallen** [comment]
> If you think about it, its a low ressource, fast & easy approach as you guys already build the transpiler to tsl : we build a shader using old way > we transpile > done

How does this not involve writing a full parser for glsl?

**Comment by Makio64** [comment]
@nkallen Maybe im missing your point but the shaderchunk system is already written ( used by the legacy webglrenderer ) and the transpiler (glsl full parser) is also already written : https://threejs.org/examples/webgpu_tsl_transpiler.html

**Comment by nkallen** [comment]
I didn't realize there was already a full glsl parser! Impressive. As for shaderchunks, I am familiar but at least for us the goal is not specifically the existing shaderchunks but our existing custom glsl shaders and custom patched shaderchunks. The parser may be able to handle all of these, I guess. At a minimum it will make converting to TSL easier.

**Comment by gkjohnson** [comment]
To drop my two cents here -

> If you would change WebGLRenderer to the same principles like the WebGL backend of WebGPURenderer, why not using WebGPURenderer in the first place?

My goal generally to get projects (libraries, large scale applications) writing with something like TSL _sooner_. Whether the project likes it or not WebGLRenderer will probably be used by a lot of projects for a long time for a lot of reasons already mentioned. All these projects are, by words used here, having to write code using the shader lib mechanism that is considered _dead_. We should be enabling and encouraging people to write code that is considered modern by the projects standards, even if that means it won't be as fully featured as the WebGPU variant. I'm very okay with TSL being "better with WebGPURenderer" as long as I can write code that's portable and modern.

To that end I'm not a huge fan of the idea of just moving WebGL's glsl code and ShaderChunk logic forward to be able to be run on WebGPU. I think this is side stepping a lot of the goals.

> I think the naming is the root cause of some frustration here

I don't agree. Whether the project has a "2" after it or not does not change the friction involved here. Applications like nkallen's will still have to upgrade whether it's called 2 or not and libraries have to support where their users are. I would expect a forward-looking transition path whether we were considering this a "version 2.0" or not.

> I still think the best way would be to revert https://github.com/mrdoob/three.js/pull/28167 and update according to the changes, there are certainly deviations that I am not calculating here, but I will give some information that may be useful.

If this is reasonably low-effort to try that would be great. I can't put time into myself and am not so deep into TSL myself, yet, but for those interested [heres a snippet of what use looked like from r160](https://github.com/mrdoob/three.js/blob/d04539a76736ff500cae883d6a38b3dd8643c548/examples/webgl_nodes_loader_gltf_iridescence.html#L80-L91) - it looks pretty transparent. It seems like it would enable what @nkallen (and I) are looking for. I'd be curious to hear @CodyJasonBennett and @RenaudRohlinger thoughts on whether this seems like a good middle ground, as well. I am supportive though. And I agree with Cody tht it would be good to confirm @Mugen87 and @mrdoob's support.

**Comment by Mugen87** [comment]
> And I agree with Cody tht it would be good to confirm @Mugen87 and @mrdoob's support.

TSL support in `WebGLRenderer` is not my personal preference but I understand it's important to honor the demands of the community. I won't block a restore of `WebGLNodeBuilder` that acts as a bridge for projects to `WebGPURenderer`. 

However, I would feel better if the project officially agrees on a single renderer strategy. Meaning _in the long-term_, `WebGLRenderer` is deprecated and we focus on `WebGPURenderer` which is hopefully renamed to `Renderer`, see #31381.

**Comment by CodyJasonBennett** [comment]
Wow. Okay, it seems like we're getting ahead of ourselves, so I will have to put code in front of you for this to not devolve into speculation. I am not suggesting anything new, so to rehash: The goal is still to enable an API like @mrdoob proposed in https://github.com/mrdoob/three.js/issues/30185#issuecomment-2638960462, which involves decoupling or implementing which renderer state `GLSLNodeBuilder` cares about. This is used for feature detection and program reflection. That allows its use with `WebGLRenderer` incidentally and is closer to the build system I would anticipate nodes to be. If it is duplicate or disjoint from `WebGPURenderer` (such as #28167), then it can be allowed to deviate for this purpose. That does create a maintenance burden that I am not sure is true should this state be relaxed. For that reason, I am not so inclined to revert the previous PR and create a parallel implementation. It is tempting though, especially if the divergence can be rectified or updated.

**Comment by sunag** [comment]
> If it is duplicate or disjoint from WebGPURenderer (such as https://github.com/mrdoob/three.js/pull/28167)

This system was never separated from the `WebGPURenderer` node system, as you can see from the PR most of the files removed were examples; there are no duplicate files in either renderer and no nodes removed either.

`WebGLNodeBuilder` does the job of routing the raw shader generated from the `NodeBuilder` into snippets that can be used in the `ShaderLib` so as not to duplicate the renderer's existing lighting model and other properties.

@mrdoob's example of `GLSLNodeBuilder` is referring to the constructor used for `WebGPURenderer` where there are already many renderer design differences. The code for `WebGLNodeBuilder` is just the shader output controller which is why it is so small, most of the code is already shared between renderers and is present in other files in the node system.

**Comment by ligaofeng0901** [comment]
I fully support the development of the WebGPURenderer. However, at the moment WebGPURenderer still lacks some features compared to the stability and maturity of WebGLRenderer. Especially when falling back to WebGL mode, we encounter several performance and stability issues.

In our case, we are working on a large-scale project with hundreds of custom shaders written in `ShaderMaterial`. We hope `WebGLRenderer` can support NodeMaterial, so that we can gradually migrate our shaders from `ShaderMaterial` to NodeMaterial within the current WebGLRenderer. This transition may take around a year.

If we had to switch directly to WebGPURenderer right now, the migration effort would be overwhelming for us. Having NodeMaterial support in WebGLRenderer would allow us to start adopting the new material system step by step in a more practical and manageable way.

**Comment by CodyJasonBennett** [comment]
> I fully support the development of the WebGPURenderer. However, at the moment WebGPURenderer still lacks some features compared to the stability and maturity of WebGLRenderer. Especially when falling back to WebGL mode, we encounter several performance and stability issues.

Open issues. These are easier to fix than porting node materials IMO, and they have to be fixed anyways.

I can make a PR with my work so far (draft based on Sunag's prior work https://github.com/mrdoob/three.js/issues/30185#issuecomment-3046493934), but a looming question is how to ensure that it stays in sync with TSL. We can trust maintainers to add stubs or notes as the language evolves, but I see this more as a quality control issue. Perhaps we can move implementation details there. Will work on getting that out later today.

**Comment by CodyJasonBennett** [comment]
It is taking longer than I hoped to update my fork, but here is another experiment vaguely similar to https://github.com/mrdoob/three.js/issues/30185#issuecomment-2638960462:

https://github.com/CodyJasonBennett/three.js/commit/745ab7446dda4078b58e59b539d4b481ea6a285a

An issue even with a dedicated `WebGLNodeBuilder` (for `WebGLRenderer`) is that `NodeBuilder` and other nodes have a dependency on the shape of the renderer, and this is used for more than just feature detection but also the actual node build process. The shim in the above experiment is not exhaustive (e.g., missing lighting/MRT). I am unsure of a pure solution or method that won't require a refactor of these nodes, as they are full of side effects and intricate dependencies. This is a design issue with nodes that will certainly complicate testing and the build process (all bundles must share references to TSL nodes; otherwise, module state is lost and nodes will crash).

**Comment by jellychen** [comment]
> It is taking longer than I hoped to update my fork, but here is another experiment vaguely similar to [#30185 (comment)](https://github.com/mrdoob/three.js/issues/30185#issuecomment-2638960462):
> 
> [CodyJasonBennett@745ab74](https://github.com/CodyJasonBennett/three.js/commit/745ab7446dda4078b58e59b539d4b481ea6a285a)
> 
> An issue even with a dedicated `WebGLNodeBuilder` (for `WebGLRenderer`) is that `NodeBuilder` and other nodes have a dependency on the shape of the renderer, and this is used for more than just feature detection but also the actual node build process. The shim in the above experiment is not exhaustive (e.g., missing lighting/MRT). I am unsure of a pure solution or method that won't require a refactor of these nodes, as they are full of side effects and intricate dependencies. This is a design issue with nodes that will certainly complicate testing and the build process (all bundles must share references to TSL nodes; otherwise, module state is lost and nodes will crash).

This is truly exciting work! Personally, I believe the node-based materials for WebGL and WebGPU don’t necessarily need to be identical. Just like Blender’s Cycles and Eevee renderers— which also have some known incompatibilities—as long as these differences are well-documented and understood, migrating between WebGL and WebGPU can become straightforward and effective.

**Comment by CodyJasonBennett** [comment]
To clarify, `WebGLRenderer` will not be as capable as `WebGLBackend` (`WebGPURenderer` with either backend) and will be best-effort. The WebGL2 feature set is very reasonable for mobile, where the usual raster pipeline is preferred to take advantage of TBDR architecture. Divergent or platform-specific code can be strategized to result in togglable features (compute/advanced post-processing) and quality loss (esp. MSAA) at the app level. That seems to be a reasonable line to draw in my experience without much effort. I do not expect to implement transform feedback or pixel buffer objects in `WebGLRenderer`, instead throwing an error, and I would strongly advise against their use and to instead write platform-specific code since WebGL easily degrades in performance with software fallbacks and expensive syncs/flushes that are difficult to orchestrate efficiently with three.js and WebGL generally. That makes device testing difficult, and I don't care to do much of that upstream in three.js without strong motivation. If I had to implement them, I would strongly suggest a duplicate or forked implementation and not to couple `WebGLRenderer` with `WebGLBackend`. I think this thread has proven that coupling is not a future-proof direction and will strain future development.

**Comment by jellychen** [comment]
> Is it technically possible to implement the backend interface in `WebGLRenderer` so it would align with `WebGLBackend`? I understand there's still a strict dependency on nodes for library/mrt fields, and node generators have a deep dependency on the current backend; otherwise, it can be used in `WebGLRenderer` all the same. Ideally, generators would not have a dependency on backends at all and reference a capabilities or pipeline layout object. I ask since I'm willing to do this work myself and would pay just for the ability to do it, even if limited to a subset of features (e.g., no transform feedback). The alternative [#29781](https://github.com/mrdoob/three.js/issues/29781) still puts the ecosystem in stasis like [pmndrs/postprocessing](https://github.com/pmndrs/postprocessing), and it's uncomfortable to do this in user-land since TSL is not a real standard, so we're beholden on a patch basis to three.js (the only implementation) all the same. In any case, I'm concerned about shutting down any stepping stone equal to any migration path.

I just tested the code you provided, and it seems it doesn't run properly. This is because GLSLNodeBuilder uses Uniform Buffer Objects (UBOs), but the current WebGLRenderer pipeline does not use UBOs to pass uniforms. As a result, the following error occurs: GL_INVALID_OPERATION: glDrawArrays: It is undefined behaviour to have a used but unbound uniform buffer.

**Comment by pailhead** [comment]
Hi, 👋 greetings from 2026!

This is a truly interesting issue, and a lot of fascinating reading.

>I don't agree. Whether the project has a "2" after it or not does not change the friction involved here

I agree with the original statement here @gkjohnson, i think naming, or versioning is indeed a source of friction. Not for everyone, but for some :) 

One exciting thing for me here is that the `WebGLRenderer` may finally become stable.

Sure, at this day and age, it's almost obsolete technology, but for a lowly peasant, that just cares about rudimentary "gpu acceleration", the `WebGLRenderer` **may suffice** and may have finally reached a stable version. 

I'm not sure in  what state of development it is currently, but if the bugs are fixed, by the virtue of people reporting them, and people still working on it,  **but no new features added**, this is a software engineers dream. Basically the version that did not see any new features for the `WebGLRenderer` is `1.0.0`, and if any fixes happened, they were patches, `1.0.12`.

I'm not sure if i read the entire thread, and what the intent is with the node builder is, ill do a second pass.

My 2 cents from what ive read are that it should definitely not be "part" of the WebGLRenderer. In fact, i would remove **ALL** references to **any and all "materials"** other than the `ShaderMaterial` and `RawShaderMaterial`. In my mind this would be a great `1.1.0` `WebGLRenderer`. 

`ShaderMaterial` is beautiful, it's awesome, and it's powerful. Most importantly, it hasn't changed, ever? The shader chunks are a bit of a different story, but they're not really part of the material, in a sense that one can write `ShaderMaterial` using only GLSL, none of the `#include <chunk>` stuff, 

I think it is possible to create an independent project, that is capable of compiling TSL into a `ShaderMaterial` thus making TSL compatible with any three version using `ShaderMaterial` alone. It's also possible to move the chunk management here, so for example, if your project is not using say a `MeshStandardMaterial` it doesn't even pull any chunks. 

FWIW, a while ago (wow almost a decade) i started playing with this idea:
https://github.com/pailhead/three-refactor-chunk-material

While this doesn't address the original issue of how to go from `ShaderMaterial` to the modern world, it could address a ton of long standing issues with three's stability and deprecation patterns. 

In short just saying:
>WebGLRenderer is now 1.0.0, the community can knock itself out building an ecosystem around it if it wants to" 

```typescript
import * as THREE from 'three';
//import { GLSLNodeBuilder, NodeMaterial } from 'three/webgpu';
import { GLSLLegacyNodeBuilder } from '@three-community/legacy-stuff'; // <--- someone is actively maintaining this
//import { color, float, time, sin, mul, add } from 'three/tsl';
import { color, float, time, sin, mul, add, NodeMaterial  } from 'three/tsl'; // <--- node material is from current three, legacy-stuff is actively following this

// community zone ================================
const builder = new GLSLLegacyNodeBuilder (); // <--community maintains this

//some modern three.js stuff
const timeFactor = sin(mul(time, float(2.0)));
const dynamicColor = mul(color(0.5, 0.2, 1.0), timeFactor);

const nodeMaterial = new NodeMaterial();
nodeMaterial.colorNode = dynamicColor;

// volounteers are targeting a stable, old, peasant WebGLRenderer
// this creates a ShaderMaterial, an API that never changed
// maybe this calls fetch to manage chunks, caching etc
const material = builder.build(nodeMaterial);

// end community zone ================================

const renderer = new THREE.WebGLRenderer();
document.body.appendChild(renderer.domElement);

const geometry = new THREE.SphereGeometry(1, 32, 32);
const mesh = new THREE.Mesh(geometry, material);  //<--- this is a ShaderMaterial
scene.add(mesh);

function animate() {
  requestAnimationFrame(animate);
  renderer.render(scene, camera); //<--- onBeforeCompile has possibly been removed, since parsing includes, ie managing chunks can happen outside of the renderer
}
animate();
```

> Applications like nkallen's will still have to upgrade whether it's called 2 or not and libraries have to support where their users are

Not to put words into anyone's mouth but the way i would then phrase this is:

> An application would still have to upgrade when it decides that the new features it needs outweighs the cost of upgrading. 

But in the meantime the existing ecosystem would be easier to work in, would be more stable etc. At some future point, the transition may be figured out such that there is no friction, but in the meantime the peasants could do peasant stuff :)

**Comment by gkjohnson** [comment]
A small update: I put some time in over the weekend to see how complicated it would be to get node materials working in WebGLRenderer and it seems like most of the necessary features are already there - only a couple small changes to WebGLRenderer were needed. It's just a matter of hooking things up. So far I've gotten meshes / points / lines, skinning, batched meshes, fog, envmaps, & lighting working with 100% TSL / node materials. Instancing and shadows may prove a bit more complicated but we'll see.

When I have a bit more time in the coming weeks I'll clean things up and make draft PR to get some feedback on the direction and how viable this feels.

| | WebGLRenderer | WebGPURenderer |
|---|---|---|
| envmap + standard node mat | <img width="733" height="546" alt="Image" src="https://github.com/user-attachments/assets/b5c38faf-8d01-402f-a2c0-36843ddae491" /> | <img width="761" height="660" alt="Image" src="https://github.com/user-attachments/assets/bbcc2fc0-360d-483a-bb1a-4bcd4300bbdc" /> |
| skinning + dir light + color node | <img height="400" alt="Image" src="https://github.com/user-attachments/assets/a78b09db-7cee-40cd-a8b9-0007c918e59f" /> | <img height="400" alt="Image" src="https://github.com/user-attachments/assets/f2004c7d-758f-47c6-a098-acf18bc4fe22" /> |
| batchedMesh + color node | <img width="702" height="716" alt="Image" src="https://github.com/user-attachments/assets/e71d6a8f-d6f6-4e9e-966f-33694b8c8140" /> | <img width="670" height="687" alt="Image" src="https://github.com/user-attachments/assets/7a004102-b33c-45ce-a928-5568468ceb7f" /> |

**Comment by gkjohnson** [comment]
I've made a PR with the compatibility layer approach in #32851 with limitations list and demos provided, as well.

**Comment by gkjohnson** [comment]
A WebGLNodesHandler class to add compatibility for TSL to WebGLRenderer has been added in #32851, designed to help aid with the transition between WebGLRenderer to WebGPURenderer. It's modeled as a dedicated class so no TSL nodes or associated classes are included unless this class is imported and added:

```js
import { WebGLRenderer } from 'three';
import { WebGLNodesHandler } from 'three/addons/tsl/WebGLNodesHandler.js';

const renderer = new WebGLRenderer();
renderer.setNodesHandler( new WebGLNodesHandler() );
```

The following examples will be available to view in the next release and are designed to test compatibility with a number of features including shadows, instancing, skinning, color node overrides, etc.

- [webgl_tsl_clearcoat](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_tsl_clearcoat.html)
- [webgl_tsl_instancing](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_tsl_instancing.html)
- [webgl_tsl_shadowmap](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_tsl_shadowmap.html)
- [webgl_tsl_skinning](https://github.com/mrdoob/three.js/blob/dev/examples/webgl_tsl_skinning.html)

There are naturally limitations, though, as both the old material system and new nodes system are not completely inter-compatible. File sizes, of course, will be bigger and some features like PMREM maps will need to be built twice if required by both material types. Other more complex built-in rendering features like VSM Shadows, transmission support, "compile" support, and other TSL-specific node features will not be supported. If there are features that are really needed they can possibly be added but hopefully this still covers a large portion of use cases and will help let projects ease into WebGPURenderer when it feels right rather than all at once.
